docs/rfcs/0007-flow-modules.md

```diff
@@ -0,0 +1,295 @@
+
+During this time the Agent team saw a lot of potential in modules outside of scraping service. Packaging up sets of components for specific workflows, publishing common use cases and allowing better usage for internal customers in the Agent as a Service model.
+
+## Goals
+
+* Support single module loading via `module.single`
+* Support multiple module loading via `module.multiple`
+* Enable re-use of common patterns
+* Allow loading a module from a string
+* Allow modules to load other modules
+* Modules should be sandboxed except via arguments and exports
+
+### Enable re-use of common patterns
+
+Common functionality can be wrapped in a set of common components that form a module. These shared modules can then be used instead of reinventing use cases. 
+
+### Allow loading a module from a string
+
+Modules will not care about the source of a string. In the case of a `module.single` the module will take in one string of valid river configuration. In the case of `module.multiple` the module will take in a `map(string)` where the key uniquely denotes the module.
+
+### Allow modules to load other modules
+
+Modules will be able to load other modules, with reasonable safe guards. There will be a stack limit for the depth of sub-modules and modules will be unable to load themselves.
+
+### Modules should be sandboxed except via arguments and exports
+
+Modules cannot directly access children or parent modules except through predefined arguments and exports. 
+
+## Non Goals
+
+* Add additional capabilities to load strings
+* Any type of versioning
+* Any user interface work beyond ensuring it works as the UI currently does
+
+*Note* should security be a non-goal? For example can a submodule spawn a `local.file`, should there be the ability to pass a deny list of modules, and/or allow list? 
+
+*Note* should metadata be a non-goal? Metadata would allow better filtering the ability to tag modules, also will allow programmatic access to things that might otherwise be in comments. 
+
+### Add additional capabilities to load strings
+
+Modules will not care about the source of the string that represents the river syntax, nor will modules have any inherent reload semantics. The component supplying the string will be responsible for the source and will notify the module when the input changes so that it can utilize the new river configuration.
+
+### Any type of versioning
+
+Modules will not contain any sort of versioning nor will check for compatibility outside the normal river checks.
+
+### Any user interface work beyond ensuring it works as the UI currently does
+
+Users will not be able to drill into modules, they will be represented as any other normal component. 
+
+## Example
+
+```river
+// module
+argument "password" {
+    optional = false
+    comment = "password for mysql"
+}
+
+argument "username" {
+    optional = false
+    comment = "username for mysql"
```

Comment:
We parse River comments into the AST, so having a comment attribute isn't necessary. We could do this:

```river 
// Username for MySQL.
argument "username" {
  optional = false
}
```

Reply:
My fear here is that since comments would not be a field it will require people knowing it exists by convention. We can document the behavior but being a field makes it more inline with how components normally work.

Reply:
Is the "comment" argument going to be read by something, or is it just meant to be a regular comment for the users who read and edit the river file? If it's used by other components/modules/UIs then I could see why it's an attribute, but if it's just a comment then I'm not sure I see the advantage vs. a regular river comment.

Reply:
It's metadata. Having all other metadata be attributes and comments not feels weird.